File: java/234_PalindromeLinkedList.java
class Solution {
    public boolean isPalindrome(ListNode head) {
        int len = getListLen(head);
        if (len == 0) {
            return true;
        }

        ListNode mid = head;
        for (int i = 0; i < (len+1) / 2; i++) {
            mid = mid.next;
        }
        mid = reverseKBack(mid);

        return isSameList(len/2, head, mid);
    }

    private int getListLen(ListNode head) {
        int len = 0;
        ListNode cur = head;
        while (cur != null) {
            cur = cur.next;
            len += 1;
        }
        return len;
    }

    private ListNode reverseKBack(ListNode head) {
        ListNode dummy = null, cur = head;
        while (cur != null) {
            ListNode next = cur.next;
            cur.next = dummy;
            dummy = cur;
            cur = next;
        }
        return dummy;
    }

    private boolean isSameList(int len, ListNode list1, ListNode list2) {
        for (int i = 0; i < len; i++) {
            if (list1.val != list2.val) {
                return false;
            }
            list1 = list1.next;
            list2 = list2.next;
        }
        return true;
    }
}

File: c++/234_PalindromeLinkedList.cpp
class Solution {
public:
    bool isPalindrome(ListNode* head) {
        int n = 0;
        for (auto p = head; p; p = p->next) n++;
        if (n <= 1) return true;
        int half = n / 2;
        auto a = head;
        for (int i = 0; i < n - half; i++) a = a->next;

        auto b = a->next;
        for (int i = 0; i < half - 1; i++) {
            auto nxt = b->next;
            b->next = a;
            a = b, b = nxt;
        }

        auto p = head, q = a;
        bool res = true;
        for (int i = 0; i < half; i++) {
            if (p->val != q->val) {
                res = false;
                break;
            }
            p = p->next, q = q->next;
        }

        auto tail = a;
        b = a->next;
        for (int i = 1; i < half - 1; i++) {
            auto nxt = b->next;
            b->next = a;
            a = b;
            b = nxt;
        }
        tail->next = nullptr;
        return res;
    }
};